google-cloud-tools-plugin/jps-plugin/src/com/google/cloud/tools/intellij/jps/GenRepoInfoFileModuleBuilder.java

```diff
@@ -0,0 +1,169 @@
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.jps.ModuleChunk;
+import org.jetbrains.jps.builders.DirtyFilesHolder;
+import org.jetbrains.jps.builders.java.JavaModuleBuildTargetType;
+import org.jetbrains.jps.builders.java.JavaSourceRootDescriptor;
+import org.jetbrains.jps.incremental.BuilderCategory;
+import org.jetbrains.jps.incremental.CompileContext;
+import org.jetbrains.jps.incremental.ModuleBuildTarget;
+import org.jetbrains.jps.incremental.ModuleLevelBuilder;
+import org.jetbrains.jps.incremental.ProjectBuildException;
+import org.jetbrains.jps.model.module.JpsModule;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collections;
+
+/**
+ * Generates source context files for a module using the Cloud SDK.
+ */
+public class GenRepoInfoFileModuleBuilder extends ModuleLevelBuilder {
+
+  private static final Logger LOG =
+      Logger.getInstance("#com.google.cloud.tools.intellij.jps.GenRepoInfoFileModuleBuilder");
+
+  public static final String NAME = "Stackdriver source context generator";
+  private GenRepoInfoFileActionFactory actionFactory = new GenRepoInfoFileActionFactory();
+
+  public GenRepoInfoFileModuleBuilder() {
+    super(BuilderCategory.CLASS_POST_PROCESSOR);
+  }
+
+  @VisibleForTesting
+  GenRepoInfoFileModuleBuilder(GenRepoInfoFileActionFactory actionFactory) {
+    this();
+    this.actionFactory = actionFactory;
+  }
+
+  /**
+   * Generates source context files for each module in {@code chunk}.
+   *
+   * <p>Generation is parameterized by the Stackdriver facet associated to the module.
+   *
+   * <p>Files are added to the same location as the project's compiled .class files.
+   */
+  @Override
+  public ExitCode build(CompileContext context, ModuleChunk chunk,
+      DirtyFilesHolder<JavaSourceRootDescriptor, ModuleBuildTarget> dirtyFilesHolder,
+      OutputConsumer outputConsumer) throws ProjectBuildException, IOException {
+    for (JpsModule jpsModule : chunk.getModules()) {
+      JpsStackdriverModuleExtension extension = jpsModule.getContainer().getChild(
+          JpsStackdriverModuleExtensionImpl.ROLE);
+
+      if (!extension.isGenerateSourceContext()) {
+        continue;
+      }
+
+      if (jpsModule.getSourceRoots().isEmpty()) {
+        LOG.warn("Module " + jpsModule.getName() + " contains no source roots. Moving on to the "
+            + "next module for source generation.");
+        continue;
+      }
+
+      if (extension.getCloudSdkPath() == null) {
+        LOG.warn("No Cloud SDK path specified. Skipping source context generation.");
+        // Cloud SDK path is a singleton for a project, so if it doesn't exist on the first module,
+        // we can return right away.
+        return ExitCode.NOTHING_DONE;
+      }
+
+      if (extension.getModuleSourceDirectory() == null) {
+        // Possible to happen on Windows, unlikely in Linux.
+        LOG.warn("Invalid module source directory. Check for special characters <>:\"|?*.");
```

Comment:
This was intended to capture the case when Paths.get() throws an `InvalidPathException`, we catch it and return null.
I don't think it's possible for getModuleSourceDirectory() to return null in any other occasion since, for new projects, it's automatically fills this field with the parent directory of the module file.